Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm

```diff
@@ -635,7 +660,7 @@ static bool isRendererThreadSafe(WebSampleBufferVideoRendering *renderering)
         if (!timebase)
             return;
 
-        purgeDecodedSampleQueue();
+        purgeDecodedSampleQueue(ShouldReschedule::No);
```

Comment:
Why do we need to purge it?
Is it an optimisation in case we have a frame with a duration, and there will be no other frame after that time.
And we want to merge the frame when the duration is passed?

Reply:
I don't actually believe we need it, in theory the head of the queue should always be the right frame for display.

This is the original code, and I did consider this potential issue:

The timerSource is always a bit late at the time its task run, When I added logging I noticed that the timebase's current time is always about 1 to 2ms late. Consistently.

Let's assume we have frame A starting at 0.000 and frame B starting at 0.017
So when the timer kicks in to purge frame A and display frame B, will actually occurs at around 0.01715 1.5ms late.
It is possible that the call to `copyDisplayedPixelBuffer` is made at 0.017 exactly, before the purge task run, we would now incorrectly retrieve frame A when B would have been expected.

Now it's probably not an issue for the majority of the case in reality as the player will not receive a notification that there's a new frame until after B frame is shown.
But the content process can call `copyDisplayedPixelBuffer` at any time.

And this is the reason why I left that call there, even when my first thought was just like yours.
> Why do we need to purge it?
Is it an optimisation in case we have a frame with a duration, and there will be no other frame after that time.
And we want to merge the frame when the duration is passed?

It's not. there's no such merge. `copyDisplayedPixelBuffer` will not retrieve the frame only if the start time of the frame is after the current time.